### PR TITLE
Usman Zaver - Scrapy

### PR DESCRIPTION
**Pull request by usmanzaver**
Scrapping Restaurants

### CODE REVIEW

File: Scrapy/yelpbot.py [hunk 1]
```diff
@@ -0,0 +1,240 @@
...
+         "/div[1]/div[4]/span[2]/span/span/a/text()")).extract()
+
+    return category
+
+
+def get_stars(response, index):
+    stars = response.xpath(
+        my_div + "/li[" + index + "]//div/@aria-label").extract()
+
+    if len(stars) == 0:
```
Comment (jawad-khan) [inline Scrapy/yelpbot.py]: or
if stars:

File: Scrapy/yelpbot.py [hunk 2]
```diff
@@ -0,0 +1,240 @@
...
+    category = response.xpath(
+        (my_div + "/li[" + index + "]/div/div/div/div[2]/div[1]/div[1]"
+         "/div[1]/div[4]/span[2]/span/span/a/text()")).extract()
+
+    return category
+
+
+def get_stars(response, index):
+    stars = response.xpath(
+        my_div + "/li[" + index + "]//div/@aria-label").extract()
```
Comment (jawad-khan) [inline Scrapy/yelpbot.py]: Use str.format.

File: Scrapy/yelpbot.py [hunk 3]
```diff
@@ -0,0 +1,240 @@
...
+
+    if len(stars) == 0:
+        pass
+    else:
+        iteration = 0
+        for list_item in stars:
```
Comment (jawad-khan) [inline Scrapy/yelpbot.py]: sounds more descriptive.
for star in stars:

File: Scrapy/yelpbot.py [hunk 4]
```diff
@@ -0,0 +1,240 @@
...
+            if "rating" in list_item:
+                total_stars = stars[iteration]
```
Comment (jawad-khan) [inline Scrapy/yelpbot.py]: you might need to rethink another efficient way.

File: Scrapy/yelpbot.py [hunk 5]
```diff
@@ -0,0 +1,240 @@
...
+    expense = response.xpath(
+        (my_div + "/li[" + index + "]/div/div/div/div[2]/div[1]/div[1]"
+         "/div[1]/div[4]/span[1]/text()")).extract()
+
+    return expense
+
+
+def get_category(response, index):
+    category = response.xpath(
+        (my_div + "/li[" + index + "]/div/div/div/div[2]/div[1]/div[1]"
```
Comment (jawad-khan) [inline Scrapy/yelpbot.py]: avoid indexing in xpaths.
use classes and ids as reference.

File: Scrapy/yelpbot.py [hunk 6]
```diff
@@ -0,0 +1,240 @@
...
+    if len(number) == 0:
+        number = response.xpath(
+            (my_div + "/li[" + index + "]/div/div/div/div[2]/div[1]"
+             "/div[2]/div/div[2]/p/text()")).extract()
+
+    return number
+
+
+def get_images(response, index):
+    image = response.xpath(
```
Comment (jawad-khan) [inline Scrapy/yelpbot.py]: is it a single image or "images"?

File: Scrapy/yelpbot.py [hunk 7]
```diff
@@ -0,0 +1,240 @@
...
+            category = get_category(response, index)
+
+            if len(category) == 0:
+                categories.append("NULL")
+            else:
+                categories.append(category)
+
+            counter += 1
+
+        else:
```
Comment (jawad-khan) [inline Scrapy/yelpbot.py]: No need for this else.

File: Scrapy/yelpbot.py [hunk 8]
```diff
@@ -0,0 +1,240 @@
...
+    stars = []
+    numbers = []
+    images = []
+    reviews = []
+    expenses = []
+    categories = []
+    addresses = []
+    hotel = None
+
+    for counter in range(1, 40):
```
Comment (jawad-khan) [inline Scrapy/yelpbot.py]: what is 40?

File: Scrapy/yelpbot.py [hunk 9]
```diff
@@ -0,0 +1,240 @@
...
+
+        hotel_no = response.xpath(
+            (my_div + "/li[" + index + "]/div/div/div/div[2]"
+             "/div[1]/div[1]/div[1]/div[1]/h3/text()")).extract_first()
+
+        if hotel_no is not None:
+            hotel = response.xpath(
+                my_div + "/li[" + index + "]//div/h3/a/text()").extract()
+
+            if len(hotel) == 0:
```
Comment (jawad-khan) [inline Scrapy/yelpbot.py]: A shorter version
if hotel:
    hotels.append(hotel)

File: Scrapy/yelpbot.py [hunk 10]
```diff
@@ -0,0 +1,228 @@
...
+            "div.lemon--div__373c0__1mboc.navigation-button-"
+            "container__373c0__2sEbf.border-color--default__373c0"
+            "__2oFDT a::attr(href)").getall()
+
+        if next_page is None:
+            print("Last page reached")
+            return
+        else:
+            if len(next_page) > 1:
+                yield response.follow(
```
Comment (jawad-khan) [inline Scrapy/yelpbot.py]: Avoid repetition of code.

File: Scrapy/settings.py [hunk 11]
```diff
@@ -0,0 +1,93 @@
+# -*- coding: utf-8 -*-
```
Comment (jawad-khan) [inline Scrapy/settings.py]: No need to include settings.py file.

File: Scrapy/yelpbot.py [hunk 12]
```diff
@@ -0,0 +1,228 @@
+import scrapy
+import csv
+
+
+'''Scrapping Restaurant information and storing it in a csv file'''
+
+DIV = str(("//*[contains(@class,'lemon--ul__373c0"
```
Comment (jawad-khan) [inline Scrapy/yelpbot.py]: Can we rename it to something relevant?

File: Scrapy/yelpbot.py [hunk 13]
```diff
@@ -0,0 +1,228 @@
...
+        if hotel_no is not None:
+            hotel = response.xpath(
+                DIV + "/li[{}]//div/h3/a/text()".format(index)).extract()
+
+            if hotel:
+                hotels.append(hotel)
+
+            address = get_address(response, index)
+
+            if address:
```
Comment (jawad-khan) [inline Scrapy/yelpbot.py]: everything related to address should be in get_address function.
same goes for other fields.

File: Scrapy/yelpbot.py [hunk 14]
```diff
@@ -0,0 +1,228 @@
...
+
+            category = get_category(response, index)
+
+            if category:
+                categories.append(category)
+            else:
+                categories.append("Not Provided")
+        counter += 1
+
+    container = []
```
Comment (jawad-khan) [inline Scrapy/yelpbot.py]: what is it?
hotels details maybe?

File: Scrapy/yelpbot.py [hunk 15]
```diff
@@ -0,0 +1,228 @@
...
+    return addresses
+
+
+def get_location(response, index):
+
+    location = response.xpath(
+        DIV + "//li[{}]/div/div/div/div[2]"
+        "/div[1]/div[2]/div/div[2]/div/p/text()".format(index)).extract()
+
+    if len(location) == 0:
```
Comment (jawad-khan) [inline Scrapy/yelpbot.py]: if location:

File: Scrapy/yelpbot.py [hunk 16]
```diff
@@ -0,0 +1,228 @@
...
+        location = response.xpath(
+            DIV + "/li[{}]/div/div/div/div[2]"
+            "/div[1]/div[2]/div/div[3]/div/p/text()".format(index)).extract()
+
+    return location
+
+
+def get_number(response, index):
+
+    number = response.xpath(
```
Comment (jawad-khan) [inline Scrapy/yelpbot.py]: Avoid generic names.

File: Scrapy/yelpbot.py [hunk 17]
```diff
@@ -0,0 +1,228 @@
...
+
+
+'''Scrapping Restaurant information and storing it in a csv file'''
+
+DIV = str(("//*[contains(@class,'lemon--ul__373c0"
+           "__1_cxs undefined list__373c0__2G8oH')]"))
+TOTAL_HOTELS = DIV + '/li/div'
+
+
+def scrapped_data(response):
```
Comment (jawad-khan) [inline Scrapy/yelpbot.py]: function name should include verb.

File: Scrapy/yelpbot.py [hunk 18]
```diff
@@ -0,0 +1,228 @@
...
+    expenses = []
+    categories = []
+    addresses = []
+    hotel = None
+    counter = 1
+    for _ in response.xpath(TOTAL_HOTELS).extract():
+        index = str(counter)
+        hotel_no = response.xpath(
+            DIV + "/li[{}]/div/div/div/div[2]"
+            "/div[1]/div[1]/div[1]/div[1]/h3"
```
Comment (jawad-khan) [inline Scrapy/yelpbot.py]: Can we avoid indexing in xpaths?

File: Scrapy/yelpbot.py [hunk 19]
```diff
@@ -0,0 +1,228 @@
...
+        ('https://www.yelp.com/search?cflt='
+         'restaurants&find_loc=San+Francisco%2C+CA')]
+         
+    def parse(self, response):
+        self.page_count += 1
+        print("procesing:" + response.url)
+        container = scrapped_data(response)
+
+        for counter in range(len(container[0])):
+            scraped_info = {
```
Comment (jawad-khan) [inline Scrapy/yelpbot.py]: can we avoid this loop?

File: Scrapy/yelpbot.py (hunk 5 above)
Comment (usmanzaver) [reply Scrapy/yelpbot.py]: I had some issues regarding this.

File: Scrapy/yelpbot.py (hunk 18 above)
Comment (usmanzaver) [reply Scrapy/yelpbot.py]: I had some issues regarding this.

File: Scrapy/yelpbot.py [hunk 20]
```diff
@@ -0,0 +1,233 @@
...
+
+
+'''Scrapping Restaurant information and storing it in a csv file'''
+
+BASE_DIV = str(("//*[contains(@class,'lemon--ul__373c0"
+                "__1_cxs undefined list__373c0__2G8oH')]"))
+TOTAL_HOTELS = BASE_DIV + '/li/div'
+
+
+class Scrapping_data():
```
Comment (jawad-khan) [inline Scrapy/yelpbot.py]: Follow class name convention from PEP8.

File: Scrapy/yelpbot.py [hunk 21]
```diff
@@ -0,0 +1,233 @@
...
+
+    def scrapping_data(self, response):
+        counter = 1
+        for _ in response.xpath(TOTAL_HOTELS).extract():
+            index = str(counter)
+
+            if self.get_hotel(response, index):
+
+                self.get_address(response, index)
+
```
Comment (jawad-khan) [inline Scrapy/yelpbot.py]: No need for extra line.

File: Scrapy/yelpbot.py [hunk 22]
```diff
@@ -0,0 +1,233 @@
...
+        hotel_no = response.xpath(
+            BASE_DIV + "/li[{}]/div/div/div/div[2]"
+            "/div[1]/div[1]/div[1]/div[1]/h3"
+            "/text()".format(index)).extract_first()
+
+        if hotel_no:
+            hotel = response.xpath(
+                BASE_DIV + "/li[{}]//div/h3/a/text()".format(index)).extract()
+        else:
```
Comment (jawad-khan) [inline Scrapy/yelpbot.py]: No need for this else

File: Scrapy/yelpbot.py [hunk 23]
```diff
@@ -0,0 +1,233 @@
...
+
+                self.get_expense(response, index)
+
+                self.get_category(response, index)
+
+            counter += 1
+
+        hotel_details = []
+        hotel_details.append(self.hotels)
+        hotel_details.append(self.addresses)
```
Comment (jawad-khan) [inline Scrapy/yelpbot.py]: Can we do this in above if condition?

File: Scrapy/yelpbot.py [hunk 24]
```diff
@@ -0,0 +1,233 @@
...
+    def get_hotel(self, response, index):
+
+        hotel_no = response.xpath(
+            BASE_DIV + "/li[{}]/div/div/div/div[2]"
+            "/div[1]/div[1]/div[1]/div[1]/h3"
+            "/text()".format(index)).extract_first()
+
+        if hotel_no:
+            hotel = response.xpath(
+                BASE_DIV + "/li[{}]//div/h3/a/text()".format(index)).extract()
```
Comment (jawad-khan) [inline Scrapy/yelpbot.py]: can we break this statement into two?

File: Scrapy/yelpbot.py [hunk 25]
```diff
@@ -0,0 +1,233 @@
...
+        if contact:
+            self.contacts.append(contact)
+        else:
+            self.contacts.append("Not Provided")
+
+        return contact
+
+    def get_image(self, response, index):
+        image = response.xpath(
+            BASE_DIV + "/li[{}]/div/div/div/div[1]/div/div/"
```
Comment (jawad-khan) [inline Scrapy/yelpbot.py]: Can we break all such kind of lines?

**Pull request merged**